github/release_notes.py

```diff
@@ -104,16 +77,94 @@ def generate_release_notes(
         commit_range = calculate_range(repository_branch, repo, helper)
     pr_numbers = fetch_pr_numbers_in_range(repo, commit_range)
     release_note_objs = fetch_release_notes_from_prs(helper, pr_numbers, current_repo_name)
-    release_notes_str = build_markdown(release_note_objs)
+    release_notes_str = MarkdownRenderer(release_note_objs).render()
 
     info(release_notes_str)
     return release_notes_str
 
-def build_markdown(
-    release_note_objs: list
-) -> str:
+class Renderer(object):
+    Node = namedtuple("Node", ["identifier", "title", "nodes", "matches_rn_field"])
+
+    target_groups = \
+        Node(
+            identifier='user',
+            title='USER',
+            nodes=None,
+            matches_rn_field='target_group_id'
+        ), \
+        Node(
+            identifier='operator',
+            title='OPERATOR',
+            nodes=None,
+            matches_rn_field='target_group_id'
+        )
+    categories = \
+        Node(
+            identifier='noteworthy',
+            title='Most notable changes',
+            nodes=target_groups,
+            matches_rn_field='category_id'
+        ), \
+        Node(
+            identifier='improvement',
+            title='Improvements',
+            nodes=target_groups,
+            matches_rn_field='category_id'
+        )
+
+    def __init__(self, release_note_objs: list):
+        self.release_note_objs = release_note_objs
+
+    def render(self)->str:
+        origin_nodes = _\
+            .chain(self.release_note_objs)\
+            .sort_by(lambda rn_obj: rn_obj.component_name.github_repo())\
+            .uniq_by(lambda rn_obj: rn_obj.source_repo)\
+            .map(lambda rn_obj: Renderer.Node(
+                identifier=rn_obj.source_repo,
+                title='[{origin_name}]'.format(origin_name=rn_obj.component_name.github_repo()),
+                nodes=Renderer.categories,
+                matches_rn_field='source_repo'
+            ))\
+            .value()
+
+        md_lines = self._nodes_to_lines(
+            nodes=origin_nodes,
+            level=1,
+            release_note_objs=self.release_note_objs
+        )
+
+        if md_lines:
+            return '\n'.join(md_lines)
+        else: # fallback
+            return 'no release notes available'
+
+    @abstractmethod
+    def _nodes_to_lines(
+        self,
+        nodes: list,
+        level: int,
+        release_note_objs: list
+    )->list:
+        pass
+
+class MarkdownRenderer(Renderer):
+
+    def _nodes_to_lines(
+        self,
+        nodes: list,
+        level: int,
+        release_note_objs: list
+    )->list:
+        lines = MarkdownRenderer._nodes_to_markdown_lines(
+            nodes=nodes,
+            level=level,
+            release_note_objs=release_note_objs
+        )
+        return lines
 
-    def get_header_suffix(
+    @staticmethod
+    def _get_header_suffix(
```

Comment:
Why not instance-methods?

Reply:
I could have, but currently the methods do not make use of instance variables as most of them are called recursive and the parameters (for the current recursion) have to be passed to the method

Reply:
Still, I'd say either define them 'public' (i.e. without the _ ) unless there is a good reason against that,  or make them instance-methods.